src/check_bgp_protocols.py

```diff
@@ -0,0 +1,141 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import os
+import re
+
+from argparse import ArgumentParser
+from subprocess import Popen, PIPE, STDOUT
+
+
+def parse_args():
+    parser = ArgumentParser(
+        description='Check if all BIRD BGP protocols are up and established'
+    )
+    parser.add_argument(
+        'path',
+        type=str,
+        help='Path to the birdc/birdc6 binary'
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    birdc_path = args.path
+
+    # Check if the path to birdc actually exists and is executable
+    if not check_birdc_executable(birdc_path):
+        print_nagios_message(
+            ExitCodes.critical,
+            'The path to the birdc executable does not exist/is not executable'
+        )
+        exit(ExitCodes.critical)
+
+    # Run the actual check and get the results
+    code, reason = check_birdc_peers(birdc_path)
+
+    print_nagios_message(code, reason)
+    exit(code)
+
+
+def check_birdc_peers(path):
+    protocols = get_birdc_protocols(path)
+
+    if not protocols:
+        return ExitCodes.critical, 'Could not retrieve BIRD\'s ' \
+                                   'output'
+
+    try:
+        r = re.compile(
+            r'^(?P<name>\w+)\s+BGP\s+(\w+)\s+(?P<state>\w+)\s+'
+            r'(?P<since>[\d\-:]+)\s+(?P<info>[\w]+)?\s*$',
+            re.MULTILINE
+        )
+
+        res = [m.groupdict() for m in r.finditer(protocols)]
+    except Exception:
+        return ExitCodes.critical, 'Could not parse BIRD\'s output'
+
+    if not res:
+        return ExitCodes.critical, 'Could not parse BIRD\'s output'
+
+    down_protocols = [protocol['name'] for protocol in res
+                      if protocol['info'] != 'Established']
+
+    if down_protocols:
+        return ExitCodes.critical, 'Some protocols are down!'
+    else:
+        return ExitCodes.ok, 'All protocols are up'
+
+
+def get_birdc_protocols(path):
+    bird = Popen(
+        [
+            path,
+            'show',
+            'protocols',
+        ],
```

Comment:
For "production" this is still absolutely readable and much less lines.

```suggestion
        [path, 'show', 'protocols'],
```